### PR TITLE
Release 4.0.0

### LINKED ISSUE #223: Proxy endpoint 400 bad request

### PR DESCRIPTION
**Pull request by SaidBySolo**
- Remove deepsource
- Change to HTTPMethodView
- Add Mirroring
- Remove Authorization

### CODE REVIEW

**Comment by SaidBySolo** [comment]
https://github.com/getsentry/sentry-python/pull/1066 must be merged and released.

File: heliotrope/database/query.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+
+async def get_all_request_count():
+    if rank_list := await RequestCount.all().values():
+        sorted_ranking = sorted(rank_list, key=lambda info: info["count"], reverse=True)
+        ranking = {
+            "total": len(sorted_ranking),
+            "list": sorted_ranking,
+        }
+        return ranking
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
        
        return {
            "total": len(sorted_ranking),
            "list": sorted_ranking,
        }
```

File: heliotrope/database/query.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        index_data.count += 1
+        await index_data.save()
+        return True
+    else:
+        if galleryinfo := await GalleryInfo.get_or_none(id=index):
+            index_data = await RequestCount.create(
+                index=index, count=1, title=galleryinfo.title
+            )
+            await index_data.save()
+            return True
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
    elif galleryinfo := await GalleryInfo.get_or_none(id=index):
        index_data = await RequestCount.create(
            index=index, count=1, title=galleryinfo.title
        )
        await index_data.save()
        return True
```

Comment (kijk2869) [inline heliotrope/database/query.py]: How about use ``create_or_none``

File: heliotrope/database/query.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        for tag_info in galleryinfo.tags:
+            tag_orm_object = Tag(
+                index_id=galleryinfo.galleryid,
+                male=tag_info.get("male"),
+                female=tag_info.get("female"),
+                tag=tag_info.get("tag"),
+                url=tag_info.get("url"),
+            )
+            await tag_orm_object.save()
+            tag_orm_object_list.append(tag_orm_object)
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
        tag_orm_object_list = list(
            map(
                    lambda tag_info: Tag(
                    index_id=galleryinfo.galleryid,
                    male=tag_info.get("male"),
                    female=tag_info.get("female"),
                    tag=tag_info.get("tag"),
                    url=tag_info.get("url"),
            ), galleryinfo.tags
            )
        )
            
        await asyncio.gather(
            *list(
                map(
                    lambda tag_orm_object: tag_orm_object.save(),
                    tag_orm_object_list
                )
            )
        )
```

File: heliotrope/database/query.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+            file_orm_object = File(
+                index_id=galleryinfo.galleryid,
+                width=file_info.get("width"),
+                hash=file_info.get("hash"),
+                haswebp=file_info.get("haswebp"),
+                name=file_info.get("name"),
+                height=file_info.get("height"),
+            )
+            await file_orm_object.save()
+            file_orm_object_list.append(file_orm_object)
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
        file_orm_object_list = list(
            map(
                lambda file_info: File(
                    index_id=galleryinfo.galleryid,
                    width=file_info.get("width"),
                    hash=file_info.get("hash"),
                    haswebp=file_info.get("haswebp"),
                    name=file_info.get("name"),
                    height=file_info.get("height"),
                ), galleryinfo.files
            )
        )
        
        await asyncio.gather(
            *list(
                map(
                    lambda file_orm_object: file_orm_object.save(),
                    file_orm_object_list
                )
            )
        )
```

File: heliotrope/database/query.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+        return [
+            {
+                **(await search_result.filter(id=search_result.id).values())[0],
+                "tags": remove_id_and_index_id(await search_result.tags.all().values()),
+                "files": remove_id_and_index_id(
+                    await search_result.files.all().values()
+                ),
+            }
+            for search_result in search_result_list
+        ]
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
        return list(map(
            lambda search_result: {
                **(await search_result.filter(id=search_result.id).values())[0],
                "tags": remove_id_and_index_id(await search_result.tags.all().values()),
                "files": remove_id_and_index_id(
                    await search_result.files.all().values()
                ),
            },
            search_result_list
        )
```

File: heliotrope/mirroring.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        self.mirroring_time = None
+        self.new_item = None
+        self.status = "idle"
+
+    async def compare_index(self):
+        remote_index_list = await self.fetch_index()
+        local_index_list = await get_index()
+
+        to_be_loaded_list = list(set(remote_index_list) - set(local_index_list))
+        return to_be_loaded_list
```
Comment (kijk2869) [inline heliotrope/mirroring.py]:
```suggestion
        return list(set(remote_index_list) - set(local_index_list))
```

File: heliotrope/routes/about.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        if is_json := request.args.get("json"):
+            if is_json in ["True", "true"]:
+                return json(
+                    {
+                        "last_checked_time": manager.last_checked_time,
+                        "last_mirrored_time": manager.mirroring_time,
+                        "new_item": manager.new_item,
+                        "server_status": manager.status,
+                    }
+                )
```
Comment (kijk2869) [inline heliotrope/routes/about.py]:
```suggestion
        if (is_json := request.args.get("json")) and (is_json.lower() == "true"):
            return json(
                {
                    "last_checked_time": manager.last_checked_time,
                    "last_mirrored_time": manager.mirroring_time,
                    "new_item": manager.new_item,
                    "server_status": manager.status,
                }
            )
```

File: heliotrope/routes/api/count.py [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+        if ranking := await get_all_request_count():
+            ranking.update({"status": 200})
+            return json(ranking)
+        return not_found
+
+    @hiyobot_only
+    async def post(self, request: HeliotropeRequest):
+        if index := request.json.get("index"):
+            if await add_request_count(index):
+                return json({"status": 200})
```
Comment (kijk2869) [inline heliotrope/routes/api/count.py:25]:
```suggestion
        if (index := request.json.get("index")) and await add_request_count(index):
            return json({"status": 200})
```

File: heliotrope/database/query.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        await galleyinfo_orm_object.tags.add(*tag_orm_object_list)
+
+
+async def put_index(index: int):
+    await Index.create(index_id=index)
+
+
+async def get_index():
+    return list(
+        map(lambda x: int(x), await Index.all().values_list("index_id", flat=True)),
```
Comment (kijk2869) [inline heliotrope/database/query.py]:
```suggestion
        map(int, await Index.all().values_list("index_id", flat=True)),
```

File: heliotrope/mirroring.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    async def start_mirroring_with_index_list(self, index_list):
+        for index in index_list:
+            if galleryinfo := await self.get_galleryinfo(index):
+                await put_galleryinfo(galleryinfo)
+
+            await put_index(index)
+        self.mirroring_time = f"({time.tzname[0]}) {datetime.datetime.now()}"
+        self.new_item = f"{len(index_list)}"
```
Comment (kijk2869) [inline heliotrope/mirroring.py]:
```suggestion
        self.new_item = str(len(index_list))
```

File: heliotrope/mirroring.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+from heliotrope.utils.requester import HitomiRequester
+
+
+class Mirroring(HitomiRequester):
+    def __init__(self, session: ClientSession) -> None:
+        super().__init__(session)
+        self.last_checked_time = None
+        self.mirroring_time = None
+        self.new_item = None
+        self.status = "idle"
```
Comment (kijk2869) [inline heliotrope/mirroring.py:18]: Make it as enum

File: heliotrope/routes/api/hitomi/images.py [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    async def get(self, request: HeliotropeRequest, index):
+        if query_galleryinfo := await get_galleryinfo(index):
+            files = HitomiImageModel.image_model_generator(query_galleryinfo["files"])
+        elif requested_galleryinfo := await request.app.ctx.hitomi_requester.get_galleryinfo(
+            index
+        ):
+            files = HitomiImageModel.image_model_generator(requested_galleryinfo.files)
+        else:
+            return not_found
+        return json(
```
Comment (kijk2869) [inline heliotrope/routes/api/hitomi/images.py:25]:
```suggestion
            return not_found
        
        return json(
```

File: heliotrope/routes/api/hitomi/images.py [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+            {
+                "files": [
+                    {
+                        "name": file.name,
+                        "image": shuffle_image_url(
+                            image_url_from_image(int(index), file, True)
+                        ),
+                    }
+                    for file in files
+                ]
```
Comment (kijk2869) [inline heliotrope/routes/api/hitomi/images.py]:
```suggestion
                "files": list(map(
                    lambda file: {
                        "name": file.name,
                        "image": shuffle_image_url(
                            image_url_from_image(int(index), file, True)
                        ),
                    }, files
                ))
```

File: heliotrope/routes/api/hitomi/search.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+from heliotrope.utils.typed import HeliotropeRequest
+
+hitomi_search = Blueprint("hitomi_search", url_prefix="/search")
+
+
+class HitomiSearchView(HTTPMethodView):
+    async def get(self, request: HeliotropeRequest):
+        if query := request.args.get("q"):
+            if result := await search_galleryinfo(query):
+                return json({"status": 200, "result": result})
```
Comment (kijk2869) [inline heliotrope/routes/api/hitomi/search.py]:
```suggestion
        if (query := request.args.get("q")) and (result := await search_galleryinfo(query)):
            return json({"status": 200, "result": result})
```

File: heliotrope/routes/api/proxy.py [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+from heliotrope.utils.shuffle import solve_shuffle_image_url
+from heliotrope.utils.typed import HeliotropeRequest
+
+heliotrope_image_proxy = Blueprint("heliotrope_image_proxy", url_prefix="/proxy")
+
+
+class HeliotropeImageProxyView(HTTPMethodView):
+    async def get(self, request: HeliotropeRequest, shuffled_url: str):
+        url = solve_shuffle_image_url(shuffled_url)
+        if not url:
```
Comment (kijk2869) [inline heliotrope/routes/api/proxy.py:14]:
```suggestion
        if not url := solve_shuffle_image_url(shuffled_url):
```

File: heliotrope/server.py [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        generate_schemas=True,
+    )
+
+
+@heliotrope_app.before_server_start
+async def start(heliotrope: Heliotrope, loop: AbstractEventLoop):
+    hitomi_session = ClientSession(loop=loop)
+    mirroring_manager_session = ClientSession(loop=loop)
+    heliotrope.ctx.hitomi_requester = HitomiRequester(hitomi_session)
+    heliotrope.ctx.mirroring_manager = Mirroring(mirroring_manager_session)
```
Comment (kijk2869) [inline heliotrope/server.py]:
```suggestion
    heliotrope.ctx.hitomi_requester = HitomiRequester(ClientSession(loop=loop))
    heliotrope.ctx.mirroring_manager = Mirroring(ClientSession(loop=loop))
```

File: heliotrope/utils/requester.py [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+#         url: StrOrURL,
+#         return_method: Literal["read", "text", "json"],
+#         **kwargs: Any,
+#     ):
+#         return await self.request("POST", url, return_method, **kwargs)
+
+
+class HitomiRequester(SessionRequester):
+    domain = "hitomi.la"
+    user_agent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36"
```
Comment (kijk2869) [inline heliotrope/utils/requester.py:113]: Use random user agent

File: heliotrope/utils/requester.py [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+    headers = {"referer": f"https://{domain}", "User-Agent": user_agent}
```
Comment (kijk2869) [inline heliotrope/utils/requester.py:114]: Use UPPER_CASE

File: heliotrope/utils/requester.py [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+                    "thumbnail": tags_model.thumbnail,
+                    "artist": tags_model.artist,
+                    "group": tags_model.group,
+                    "type": tags_model.hitomi_type,
+                    "language": tags_model.language,
+                    "series": tags_model.series,
+                    "characters": tags_model.characters,
+                    "tags": tags_model.tags,
+                }
+                return tags_dict
```
Comment (kijk2869) [inline heliotrope/utils/requester.py]:
```suggestion
                return {
                    "title": tags_model.title,
                    "thumbnail": tags_model.thumbnail,
                    "artist": tags_model.artist,
                    "group": tags_model.group,
                    "type": tags_model.hitomi_type,
                    "language": tags_model.language,
                    "series": tags_model.series,
                    "characters": tags_model.characters,
                    "tags": tags_model.tags,
                }
```

File: heliotrope/utils/shuffle.py [hunk 20]
```diff
@@ -18,13 +22,13 @@ def shuffle_image_url(url: str) -> str:
...
 
-def solve_shuffle_image_url(shuffled_image_url: str) -> str:
+def solve_shuffle_image_url(shuffled_image_url: str) -> Optional[str]:
     try:
         solve_regex: list[str] = re.findall(
             r"(.+?)_(.+?)_(.+?_net|.+?la)_(.+)_(.+?_.+)", shuffled_image_url
         )[0]
     except:
-        return json({"code": 400, "message": "bad_request"}, 400)
+        return
```
Comment (kijk2869) [inline heliotrope/utils/shuffle.py]: Use contextlib.suppress

File: heliotrope/utils/shuffle.py [hunk 21]
```diff
@@ -1,10 +1,14 @@
...
-
-def shuffle_image_url(url: str) -> str:
-    url_parse_regex = re.compile(r"\/\/(..?)(\.hitomi\.la|\.pximg\.net)\/(.+?)\/(.+)")
+def shuffle_image_url(url: str) -> Optional[str]:
+    try:
+        url_parse_regex = re.compile(
+            r"\/\/(..?)(\.hitomi\.la|\.pximg\.net)\/(.+?)\/(.+)"
+        )
+    except:
+        return
```
Comment (kijk2869) [inline heliotrope/utils/shuffle.py]: Use contextlib.suppress

File: heliotrope/routes/api/hitomi/list.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+hitomi_list = Blueprint("hitomi_list", url_prefix="/list")
+
+
+class HitomiListView(HTTPMethodView):
+    async def get(self, request: HeliotropeRequest, index: int):
+        hitomi_index_list = await request.app.ctx.hitomi_requester.fetch_index()
+        split_hitomi_index_list = [
+            hitomi_index_list[i * 15 : (i + 1) * 15]
+            for i in range((len(hitomi_index_list) + 15 - 1) // 15)
+        ]
```
Comment (kijk2869) [inline heliotrope/routes/api/hitomi/list.py]:
```suggestion
        split_hitomi_index_list = list(map(
            lambda i: hitomi_index_list[i * 15 : (i + 1) * 15], 
            range((len(hitomi_index_list) + 15 - 1) // 15)
        ))
```

File: heliotrope/routes/api/hitomi/list.py [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+        start_at_zero = index - 1
+
+        if len(split_hitomi_index_list) < start_at_zero + 1 or start_at_zero < 0:
+            return not_found
+
+        info_list = await gather(
+            *[
+                request.app.ctx.hitomi_requester.get_info_using_index(index)
+                for index in split_hitomi_index_list[index]
+            ]
```
Comment (kijk2869) [inline heliotrope/routes/api/hitomi/list.py]:
```suggestion
            *list(map(
                lambda index: request.app.ctx.hitomi_requester.get_info_using_index(index), 
                split_hitomi_index_list[index]
            ))
```

File: heliotrope/utils/hitomi/models.py [hunk 24]
```diff
@@ -0,0 +1,170 @@
...
+        self.title = title
+        self.galleryid = galleryid
+        self.hitomi_type = hitomi_type
+
+    @classmethod
+    def parse_galleryinfo(cls, galleryinfo_json, parse: bool = False):
+        if not galleryinfo_json["tags"]:
+            parsed_tags = []
+        else:
+            parsed_tags = []
```
Comment (kijk2869) [inline heliotrope/utils/hitomi/models.py]:
```suggestion
        parsed_tags = []
        
        if galleryinfo_json["tags"]:
```

File: heliotrope/utils/hitomi/models.py [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+                        {"value": f"male:{tag['tag']}", "url": tag["url"]}
+                    )
+                elif not tag.get("male") and not tag.get("female"):
+                    parsed_tags.append(
+                        {"value": f"tag:{tag['tag']}", "url": tag["url"]}
+                    )
+                elif tag.get("male") and tag.get("female"):
+                    raise Exception
+                else:
+                    raise Exception
```
Comment (kijk2869) [inline heliotrope/utils/hitomi/models.py]:
```suggestion
                if len(value := list(filter(None, ["male" if tag.get("male") else None, "female" if tag.get("female") else None]))) > 1:
                    raise Exception

            parsed_tags.append({"value": f"{value[0] if value else 'tag'}:{tag['tag']}", "url": tag["url"]})
```

File: heliotrope/utils/hitomi/models.py [hunk 26]
```diff
@@ -0,0 +1,170 @@
...
+        self.language = language
+        self.series = series
+        self.characters = characters
+        self.tags = tags
+
+    @staticmethod
+    def check_element(elements):
+        if isinstance(elements, list):
+            if not elements:
+                return []
```
Comment (kijk2869) [inline heliotrope/utils/hitomi/models.py:121]: Should it returns an empty list when elements is empty list?

**Review by kijk2869 — CHANGES_REQUESTED** [review]

File: heliotrope/utils/hitomi/models.py (hunk 26 above)
Comment (kijk2869) [reply heliotrope/utils/hitomi/models.py:121]: If not, combine those two if comprehensions.

File: heliotrope/routes/about.py (hunk 7 above)
Comment (SaidBySolo) [reply heliotrope/routes/about.py]: Syntax error
```suggestion
        if is_json := request.args.get("json"):
            if is_json.lower() == "true":
                return json(
                    {
                        "last_checked_time": manager.last_checked_time,
                        "last_mirrored_time": manager.mirroring_time,
                        "new_item": manager.new_item,
                        "server_status": manager.status,
                    }
                )
```

File: heliotrope/routes/api/hitomi/images.py (hunk 12 above)
Comment (SaidBySolo) [reply heliotrope/routes/api/hitomi/images.py:25]: files possibly unbound

File: heliotrope/routes/api/hitomi/search.py (hunk 14 above)
Comment (SaidBySolo) [reply heliotrope/routes/api/hitomi/search.py]: Syntax Error

Comment (kijk2869) [reply heliotrope/routes/api/hitomi/search.py]: I edited it

File: heliotrope/routes/about.py (hunk 7 above)
Comment (kijk2869) [reply heliotrope/routes/about.py]: > Syntax error
> ```suggestion
>         if is_json := request.args.get("json"):
>             if is_json.lower() == "true":
>                 return json(
>                     {
>                         "last_checked_time": manager.last_checked_time,
>                         "last_mirrored_time": manager.mirroring_time,
>                         "new_item": manager.new_item,
>                         "server_status": manager.status,
>                     }
>                 )
> ```
> 

Do not use this, i edited it.

**Pull request merged**